src/std/misc/repr.ss

```diff
@@ -0,0 +1,130 @@
+;; The port defaults to (current-output-port). The options are reserved for future use.
+;; : <- Any (Optional Port) (Optional Table)
+(def (print-representation x (port (current-output-port)) (options (current-representation-options)))
+
+  ;; Our universal utilities: print (recurse), repr, display, write
+  (def (p y (q port)) (print-representation y q options))
+  (def (r y) (repr y options))
+  (def (d y) (display y port))
+  (def (w y) (write y port))
+  (def (simple? x) (or (number? x) (boolean? x) (string? x) (char? x) (void? x) (eof-object? x)))
+
+  (cond
+   ;; Simplest case: just write it.
+   ((simple? x)
+    (w x))
+   ((or (symbol? x) (null? x)) ;; requires slightly more care: write it after a quote.
+    (d "'") (w x))
+   ((pair? x) ;; pair: print as [ ... ].
+    (display-separated x prefix: "[" separator: " " display-element: p port: port)
+    (let ((end (cdr (last-pair x))))
+      (cond
+       ((null? end) (void))
+       ((simple? end) (d " . ") (p end))
+       (else (d " ") (p end) (d " ..."))))
+    (d "]"))
+   ((vector? x) ;; vector: print as (vector ...).
+    (display-separated (vector->list x)
+                       prefix: "(vector " separator: " " display-element: p suffix: ")" port: port))
+   ((u8vector? x) ;; u8vector: print as #u8(...), knowing that elements are all bytes.
+    (display-separated (u8vector->list x)
+                       prefix: "#u8(" separator: " " display-element: d suffix: ")" port: port))
+   ((hash-table? x) ;; hash-table: print as (hash ...)
+    ;; NB: also assumes (1) it is a equal? table, and (2) you use :std/sugar ...
+    (display-separated
+     (sort (map (lambda (k) (cons (r k) (hash-ref x k))) (hash-keys x)) ;; sort keys by repr
+           (lambda (krv1 krv2) (string<? (car krv1) (car krv2))))
+     port: port prefix: "(hash " suffix: ")"
+     display-element: (lambda (x _) (match x ([kr . v] (d "(") (d kr) (d " ") (p v) (d ")"))))))
+   ((representable? x)
+    {:pr x port options})
+   ((and (object? x) (find-method (object-type x) ':pr))
+    => (lambda (m) (m x port options)))
+   ((and (object? x) (struct-type? (object-type x)))
+    (display-separated
+     (cdr (struct->list x))
+     port: port prefix: (format "(~a " (##type-name (object-type x))) suffix: ")"
+     display-element: p))
+   (else
+    (print-unrepresentable-object x port options))))
+
+;; Given an object without a known evaluable representation, a port and options,
+;; print a general-purpose escape, using the #42 syntax and putting in a string
+;; a hint as to what the value is, as obtained from the write function.
+;; : <- Any Port Table
+(def (print-unrepresentable-object
+      object (port (current-output-port)) (options (current-representation-options)))
+  (def (d x) (display x port))
+  (def (w x) (write x port))
+  (d "(begin0 #") (d (object->serial-number object)) (d " ") (w (object->string object)) (d ")"))
```

Comment:
i think this should be just object->string, that representation is unreadable already.

Reply:
No. We *really* want the form to be evaluatable no matter what. `(begin0 #34 "#<procedure #34 foo>")` is useful in a wide variety of settings.

Reply:
ok -- either way works.